wallet-new/src/Cardano/Wallet/Kernel/DB/Sqlite.hs

```diff
@@ -0,0 +1,562 @@
+
+--}
+data TxMetaT f = TxMeta {
+      _txMetaTableId         :: Beam.Columnar f Core.TxId
+    , _txMetaTableAmount     :: Beam.Columnar f Core.Coin
+    , _txMetaTableCreatedAt  :: Beam.Columnar f Core.Timestamp
+    , _txMetaTableIsLocal    :: Beam.Columnar f Bool
+    , _txMetaTableIsOutgoing :: Beam.Columnar f Bool
+    } deriving Generic
+
+type TxMeta = TxMetaT Identity
+
+deriving instance Eq TxMeta
+deriving instance Show TxMeta
+
+-- | Creates a storage-specific 'TxMeta' out of a 'Kernel.TxMeta'.
+mkTxMeta :: Kernel.TxMeta -> TxMeta
+mkTxMeta txMeta = TxMeta {
+                  _txMetaTableId         = txMeta ^. Kernel.txMetaId
+                , _txMetaTableAmount     = txMeta ^. Kernel.txMetaAmount
+                , _txMetaTableCreatedAt  = txMeta ^. Kernel.txMetaCreationAt
+                , _txMetaTableIsLocal    = txMeta ^. Kernel.txMetaIsLocal
+                , _txMetaTableIsOutgoing = txMeta ^. Kernel.txMetaIsOutgoing
+                }
+
+instance Beamable TxMetaT
+
+instance Table TxMetaT where
+    data PrimaryKey TxMetaT f = TxIdPrimKey (Beam.Columnar f Core.TxId) deriving Generic
+    primaryKey = TxIdPrimKey . _txMetaTableId
+
+instance Beamable (PrimaryKey TxMetaT)
+
+{--
+
+* Table 2: ~tx_meta_inputs~
+** Primary Index: ~tx_meta_input_address~
+** Secondary Indexes: ~tx_meta_id~
+
+| tx_meta_input_address | tx_meta_coin | tx_meta_id |
+|-----------------------+--------------+------------|
+| Core.Address          | Core.Coin    | Core.TxId  |
+
+** Table 3: ~tx_meta_outputs~
+** Primary Index: ~tx_meta_output_address~
+** Secondary Indexes: ~tx_meta_id~
+
+| tx_meta_output_address | tx_meta_coin | tx_meta_id |
+|------------------------+--------------+------------|
+| Core.Address           | Core.Coin    | Core.TxId  |
+
+--}
+
+data TxCoinDistributionTableT f = TxCoinDistributionTable {
+      _txCoinDistributionTableAddress :: Beam.Columnar f Core.Address
+    , _txCoinDistributionTableCoin    :: Beam.Columnar f Core.Coin
+    , _txCoinDistributionTxId         :: Beam.PrimaryKey TxMetaT f
+    } deriving Generic
+
+type TxCoinDistributionTable = TxCoinDistributionTableT Identity
+
+instance Prelude.Show TxCoinDistributionTable where
+    show txCoin =
+      let (TxIdPrimKey txid) = _txCoinDistributionTxId txCoin
+      in formatToString ("id = " % build % ", address = " % build % ", coin = " % build)
+                        txid
+                        (_txCoinDistributionTableAddress txCoin)
+                        (_txCoinDistributionTableCoin txCoin)
+
+instance Beamable TxCoinDistributionTableT
+
+-- | The inputs' table.
+newtype TxInputT f = TxInput  {
+  _getTxInput  :: (TxCoinDistributionTableT f)
+  } deriving Generic
+
+type TxInput = TxInputT Identity
+deriving instance Show TxInput
+
+-- | Generalisation of 'mkInputs' and 'mkOutputs'.
+mkCoinDistribution :: forall a. Kernel.TxMeta
+                   -> (Getter Kernel.TxMeta (NonEmpty (Core.Address, Core.Coin)))
```

Comment:
I'm inclined to say one almost never wants a `Getter s a` as an argument, rather than an `s -> a`. One can use `to (_ :: s -> a) :: Getter s a` if necessary.